File: go/update.go
package immgo

func Update(hostTree HostTree, hostRoot HostNode, shadowRoot *ShadowNode, render RenderFunc2) error {

	renderRoot := NewRenderNode("root", "root", Properties{})
	renderRoot.data.match = shadowRoot

	// trigger at most one event
	err := hostTree.TriggerEvent()
	if err != nil {
		return err
	}

	// render
	render(renderRoot)

	// commit
	hostTree.BeginFrame()
	err = Commit(renderRoot, hostRoot, hostTree)
	if err != nil {
		return err
	}

	hostTree.EndFrame()
	return nil
}

File: py/immpy/update.py
from typing import Callable

from renderer import Renderer
from host_tree import HostTree, HostNode
from node import ShadowNode

RenderFunc = Callable[[Renderer], None]

def update(host_tree: HostTree, host_node: HostNode, shadow_node: ShadowNode,
        render: RenderFunc):
    host_tree.trigger_event()

    renderer = Renderer(shadow_node)
    render(renderer)
    
    host_tree.begin_frame()
    renderer.commit(host_tree, host_node)
    host_tree.end_frame()
